[SIL] Introduce createStringLiteralWithNullTerminator

vedantk opened the pull request:
This patch changes the behavior of `createStringLiteral` so that it does not null-terminate strings. It introduces `createStringLiteralWithNullTerminator` to ease the transition, as discussed in PR #1180.

Since this required a change to `createStringConstant`, I expect to see some test failures. I don't have the resources to test this patch locally so I'm relying on the CI to catch these.

vedantk committed:
[SIL] Introduce createStringLiteralWithNullTerminator

vedantk committed:
[IRGen] Do not implicitly nul terminate strings

vedantk commented:
@swift-ci Please test

dabrahams commented:
is this a good idea?  From the description it sounds like it will cause new allocation and copying when a string literal is passed to a C API.

vedantk commented:
I don't think this patch introduces any extra allocations. Actually, I expect the number of heap allocations to decrease.

The current behavior of `createStringConstant` may cause an extra allocation in llvm (since AddNull defaults to true, see http://llvm.org/docs/doxygen/html/Constants_8cpp_source.html#l02552). We get rid of the possible allocation there by paying for it up front when needed in `createStringLiteralWithNullTerminator`.

vedantk commented:
But you're right, the plan of record is to get rid of `createStringLiteralWithNullTerminator` ASAP. I'm open to any way of creating a StringLiteralInst _without_ appending a null terminator so PR #1239 can make progress.

vedantk commented:
@swift-ci Please test

vedantk committed:
Remove nul terminators more conservatively

vedantk committed:
Thread addNull through getAddrOfGlobalString

vedantk committed:
Null terminate empty strings

vedantk commented:
@swift-ci Please test

vedantk committed:
Fix createStringConstant invocation

vedantk commented:
@swift-ci Please test

vedantk committed:
Make null check utf8-exclusive

vedantk commented:
@swift-ci Please test

vedantk committed:
Actually disable appending null terminators

vedantk commented:
@swift-ci Please test

vedantk committed:
Fix up two affected test cases

vedantk commented:
@jckarter @dabrahams All tests passing now :). To summarize, the key change here is changing `getAddrOfString` in IRGenSIL.cpp to _not_ append null terminators to all strings. Does the patch seem reasonable?

jckarter commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2260,9 +2260,11 @@ void IRGenSILFunction::visitFloatLiteralInst(swift::FloatLiteralInst *i) {
+  case swift::StringLiteralInst::Encoding::UTF8: {
+    return IGM.getAddrOfGlobalString(string,
+                                     /*willBeRelativelyAddressed=*/false,
+                                     /*addNull=*/false);
```
Can we remove the `addNull` parameter now?

vedantk replied:
Not quite yet. Several parts of Swift still rely on `getAddrOfGlobalString` to null-terminate strings.

jckarter replied:
I think that's also better handled by adding a `getAddrOfNullTerminatedGlobalString` helper that adds the null terminator then calls `getAddrOfGlobalString`. As written, you're caching only keyed by the input `StringRef`, so this would break:

```
getAddrOfGlobalString("foo", false, false);
getAddrOfGlobalString("foo", false, true); // will return unterminated 'foo' from cache
```

vedantk committed:
Remove the addNull parameter from IGM.getAddrOfGlobalString

vedantk committed:
Remove the addNull parameter from createStringConstant

vedantk commented:
@swift-ci Please test

vedantk committed:
Fix typo

vedantk commented:
@swift-ci Please test

vedantk committed:
Try harder to maintain existing behavior (no spurious test updates)

Also, address review comments.

vedantk committed:
Fix merge conflict with master

vedantk committed:
Fix more merge conflicts

vedantk committed:
De-dup code in SILBuilder, use tests from master

vedantk committed:
Remove another duplicated symbol

vedantk commented:
@swift-ci Please test

vedantk committed:
Get rid of an extra copy

vedantk commented:
@jckarter I think this version addresses your review comments. We cache string literals properly now, there are separate 'null terminated' variants of `createConstantString` and `getAddrOfGlobalString`, and the patch is strictly NFC (no tests are updated). Wdyt?

jckarter commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -1222,12 +1222,14 @@ class StringLiteralInst final : public LiteralInst,
 
   StringLiteralInst(SILDebugLocation *DebugLoc, StringRef text,
-                    Encoding encoding, SILType ty);
+                    Encoding encoding, bool nullTerminate, SILType ty);
```
`nullTerminate` doesn't need to be a flag here. We should pass down a StringRef including the null terminator.

vedantk replied:
That approach breaks a few tests which require literals to be null terminated, but which assert-fail in llvm if a null is actually included in the StringRef.

Getting rid of this flag makes SLI->getValue() more complicated, and forces the frontend to make copies of string literals to ensure that they are null-terminated.

jckarter replied:
Or, at least, allocate the null terminator into the StringLiteralInst's tail buffer, so we don't need to check this extra flag everywhere.

jckarter replied:
Can you be specific as to what breaks? You could still avoid the caller having to copy by having the constructor take a `Twine` instead, maybe, or take the flag and then append the null terminator in the `create`-d buffer that gets allocated.

jckarter replied:
`SLI->getValue()` should include the null terminator if that's part of the value.

vedantk replied:
This `isNullTerminated` flag only needs to be checked when lowering a SLI to an llvm constant, which will only ever happen in exactly one place (`getAddrOfString`).

If `SLI->getValue()` includes the null terminator when it's part of the final value, 5 test cases start failing (I think one dealt with passing down ObjC identifiers to llvm).

jckarter replied:
We should be able to fix that. Which tests?

jckarter replied:
I see 12 uses of `SLI::getValue()` in Xcode's index, and all of them look like they should consider the null terminator.

DougGregor commented:
@vedantk , @jckarter this is a very old PR. Is it still relevant?

vedantk commented:
No, I'll go through any other's I've got and clean them up.
